File: runtime/runtime-impl/src/main/python/sga_runtime/record.py
from abc import ABC, abstractmethod
from typing import Any, List, Tuple


class AbstractRecord(ABC):
    @abstractmethod
    def key(self):
        pass

    @abstractmethod
    def value(self):
        pass

    @abstractmethod
    def origin(self) -> str:
        pass

    @abstractmethod
    def timestamp(self) -> int:
        pass

    @abstractmethod
    def headers(self) -> List[Tuple[str, Any]]:
        pass


class Record(AbstractRecord):
    def __init__(self, value, key=None, origin: str = None, timestamp: int = None,
                 headers: List[Tuple[str, Any]] = None):
        self._value = value
        self._key = key
        self._origin = origin
        self._timestamp = timestamp
        self._headers = headers or []

    def key(self):
        return self._key

    def value(self):
        return self._value

    def origin(self) -> str:
        return self._origin

    def timestamp(self) -> int:
        return self._timestamp

    def headers(self) -> List[Tuple[str, Any]]:
        return self._headers

    def __str__(self):
        return f"Record(value={self._value}, key={self._key}, origin={self._origin}, timestamp={self._timestamp}, " \
               f"headers={self._headers})"

File: langstream-api/src/main/java/ai/langstream/api/runner/code/Record.java
/*
 * Copyright DataStax, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package ai.langstream.api.runner.code;

import java.util.Collection;

public interface Record {
    Object key();

    Object value();

    String origin();

    Long timestamp();

    Collection<Header> headers();

    default Header getHeader(String key) {
        return headers().stream().filter(h -> h.key().equals(key)).findFirst().orElse(null);
    }
}
